src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxy.cs

```diff
@@ -39,5 +39,35 @@ protected DispatchProxy()
         {
             return (T)DispatchProxyGenerator.CreateProxyInstance(typeof(TProxy), typeof(T));
         }
+
+        /// <summary>
+        /// Creates an object instance that derives from class <paramref name="proxyType"/>
+        /// and implements interface <paramref name="interfaceType"/>.
+        /// </summary>
+        /// <param name="interfaceType">The interface the proxy should implement.</param>
+        /// <param name="proxyType">The base class to use for the proxy class.</param>
+        /// <returns>An object instance that implements <paramref name="interfaceType"/>.</returns>
+        /// <exception cref="System.ArgumentException"><paramref name="interfaceType"/> is a class,
+        /// or <paramref name="proxyType"/> is sealed or does not have a parameterless constructor</exception>
+        [RequiresDynamicCode("Creating a proxy instance requires generating code at runtime")]
+        public static object Create([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type interfaceType, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type proxyType)
+        {
+            if (interfaceType == null)
+            {
+                throw new ArgumentNullException(nameof(interfaceType));
+            }
+
+            if (proxyType == null)
+            {
+                throw new ArgumentNullException(nameof(proxyType));
+            }
+
+            if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))
+            {
+                throw new ArgumentException(SR.Format(SR.ProxyType_Must_Be_Assignable_To_DispatchProxy, proxyType.Name));
+            }
```

Comment:
This will need more validation - many kinds of bad Type instances can be passed to this overload. For example:

```csharp
// Passes all the validations so far (prints true)
Console.WriteLine(typeof(BadProxy1<>).IsAssignableTo(typeof(DispatchProxy)));
// Passes all the validations so far (prints true)
Console.WriteLine(typeof(BadProxy2<>).GetGenericArguments()[0].IsAssignableTo(typeof(DispatchProxy)));

abstract class BadProxy1<T> : DispatchProxy { }
class BadProxy2<T> where T : DispatchProxy { }
```

We also need to define how this should react if someone passes a Type instance that is not even owned by the runtime. E.g. if it's obtained from a MetadataLoadContext.

Reply:
> This will need more validation - many kinds of bad Type instances can be passed to this overload.

Added test scenarios you mentioned, could not find other bad cases, it seems existing tests has quite good coverage.
https://github.com/dotnet/runtime/blob/a4bc7fab9cf9628d8ab7cd1b12e6ab2516ce9046/src/libraries/System.Reflection.DispatchProxy/tests/DispatchProxyTests.cs#L136-L177
The problem here: the argument names passed to ArgumentException is same as generic overload type parameter name, I am thinking to pass the argument names where the exceptions thrown



> We also need to define how this should react if someone passes a Type instance that is not even owned by the runtime. E.g. if it's obtained from a MetadataLoadContext.

I think we should throw an ArgumentException in that case, not sure how to check if a type is loaded by MLC, added a test, it fails on `if (!proxyType.IsAssignableTo(typeof(DispatchProxy)))` condition and throws

Reply:
> We also need to define how this should react if someone passes a Type instance that is not even owned by the runtime. E.g. if it's obtained from a MetadataLoadContext.

If a type is owned by MetadataLoadContext then its base class (i.e. `DispatchProxy`) will also be a separate type from the runtime `DispatchProxy` type and thus doesn't need to be specially checked since `IsAssignableTo` will return `false`.

Reply:
If we need it, I suppose a check like `typeof(object).GetType() == proxyType.GetType()` would differentiate between RuntimeType and others (RoType).

Reply:
It is not safe to assume that all runtime type objects are of the same type.

Reply:
`proxyType.Assembly.ReflectionOnly` works for existing tests, but not sure if `Assembly.ReflectionOnly` would always work as ReflectionOnly loading is deprecated

Reply:
To recap where I think we are here:

There are 2 Types to validate: `proxyType` and `interfaceType`.

For `proxyType` should just be able to do `proxyType.IsAssignableTo(typeof(DispatchProxy))` which should work fine with MLC and other cases where the base Type system is custom.

For `interfaceType`, if we don't validate it, we will eventually get an exception because there is no `Type.TypeHandle`. So if we still want to validate this, we should use an established pattern assuming there is one and that we've validated types before in similar cases. 

@jkotas do you know what the pattern is to validate that a Type is a valid runtime type? Thoughts:
- Use `type.TypeHandle` and then validate a non-default value. MLC throws InvalidOperation so we probably just let it throw without try\catch.
- Use `type is RuntimeType` or `type.TypeHandle is RuntimeTypeHandle`. However, since those types are internal only works for internal\core code unless we find\add a wrapper. 
- Use `typeof(object).GetType() == type.GetType()`. May not work in all cases?
- Use `proxyType.Assembly.ReflectionOnly` (probably not since deprecated)

Reply:
Thanks for the recap, that is exactly what I need

Reply:
I do not think we need to bother checking that the passed in type is RuntimeType. We do not do it anywhere else in the libraries and just let it throw exception from within the implementation if it calls unsupported method. For example, try the following with System.Text.Json:
```csharp
using System.Reflection;
using System.Reflection.Emit;
using System.Text.Json;

AssemblyBuilder ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("MyAssembly"), AssemblyBuilderAccess.Run);
ModuleBuilder mb = ab.DefineDynamicModule("MyModule");
TypeBuilder tb = mb.DefineType("MyType", TypeAttributes.Public);

JsonSerializer.Deserialize("{ }", tb);
```
It is going to throw at:
```
Unhandled exception. System.NotSupportedException: The invoked member is not supported before the type is created.
   at System.Reflection.Emit.TypeBuilder.GetCustomAttributes(Type attributeType, Boolean inherit)
   at System.Text.Json.Reflection.ReflectionExtensions.GetUniqueCustomAttribute[TAttribute](MemberInfo memberInfo, Boolean inherit)
   at System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.GetConverterForType(Type typeToConvert, JsonSerializerOptions options, Boolean resolveJsonConverterAttribute)
   at System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.CreateJsonTypeInfo(Type type, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.GetTypeInfo(Type type, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializerOptions.GetTypeInfoNoCaching(Type type)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at System.Text.Json.JsonSerializerOptions.GetTypeInfoInternal(Type type, Boolean ensureConfigured, Boolean resolveIfMutable)
   at System.Text.Json.JsonSerializer.GetTypeInfo(JsonSerializerOptions options, Type inputType)
   at System.Text.Json.JsonSerializer.Deserialize(String json, Type returnType, JsonSerializerOptions options)
   at Program.<Main>$(String[] args) in C:\repro\Program.cs:line 9
```
The reflection object model has been (at least partially) designed to allow custom types pretending to be runtime types so allowing custom types to be passed in is not necessarily wrong.

Reply:
Some comments on the proposed options:

> Use type.TypeHandle and then validate a non-default value. MLC throws InvalidOperation so we probably just let it throw without try\catch.
> Use proxyType.Assembly.ReflectionOnly (probably not since deprecated)

The custom types can override these methods and return whatever from them, so this would be assuming that custom type is implemented correctly. You can as well assume that the custom type is implemented correctly to make everything else work.

> Use type is RuntimeType or type.TypeHandle is RuntimeTypeHandle. However, since those types are internal only works for internal\core code unless we find\add a wrapper.

Right, you would need to have access to runtime internals.

> Use typeof(object).GetType() == type.GetType(). May not work in all cases?

This depends on the runtime implementain details. It happens to work for the current CoreCLR reflection implementation, but it won't work for current native AOT reflection implementation.

Reply:
> I do not think we need to bother checking that the passed in type is RuntimeType. We do not do it anywhere else in the libraries and just let it throw exception from within the implementation if it calls unsupported method.

Sounds good to me thanks, will remove new test with MLC